lib/cloud_controller/permissions.rb

```diff
@@ -237,6 +241,14 @@ def can_read_route?(space_id)
       @user.audited_organizations.map(&:id).include?(space.organization_id)
   end
 
+  def space_guids_with_readable_routes_query
+    if can_read_globally?
+      VCAP::CloudController::SPACE.select(:guid)
```

Comment:
This line should never be reached. Otherwise it would be very expensive to first select all `space_guids` and then use this giant list in some filter. Thus I would recommend to raise an exception like we did e.g. [here](https://github.com/cloudfoundry/cloud_controller_ng/blob/8173f4677b350b79dc5c985a3005119b9644919a/lib/cloud_controller/permissions.rb#L157).

Reply:
I am surprised to hear that calling `space_guids_with_readable_routes_query` as an admin would lead to expensive queries. I naively assumed that it was safe because it returns a dataset rather than an in memory object. 

I switched to throwing an error because I know y'all have been doing a lot of work to optimize performance but would love to hear your thoughts.

Reply:
The call to `space_guids_with_readable_routes_query` as admin itself would not be expensive, as no DB query would be executed, but the resulting dataset could then be used by the caller in some other DB query that will be executed. And returning all space guids and using them in another WHERE condition just does not make sense, i.e. `SELECT * FROM apps WHERE space_guid IN (SELECT guid FROM spaces)` is semantically the same as `SELECT * FROM apps`.

Thus it's always better to support the two different scenarios (i.e. admin vs. non-admin) directly in the DB query that's going to be executed (as you did [here](https://github.com/cloudfoundry/cloud_controller_ng/blob/d58cb6afb8fa90f14bfb3205e33c7a305c50397a/app/fetchers/route_fetcher.rb#L11-L16).

The exception should only be there to ensure that no one accidentally forgets this when adding some new functionality.